main/src/ca/uwaterloo/flix/language/phase/Safety.scala

```diff
@@ -33,303 +33,406 @@ object Safety {
+        val rulesVal = traverse(rules) { case HandlerRule(op, fparams, exp) =>
+          mapN(visit(exp)) { e => HandlerRule(op, fparams, e) } }
+        mapN(expVal, rulesVal) { (e, rs) => Expression.TryWith(e, effUse, rs, tpe, pur, eff, loc) }
+
+      case Expression.Do(op, exps, pur, eff, loc) =>
+        val expsVal = traverse(exps)(visit)
+        mapN(expsVal) { es => Expression.Do(op, es, pur, eff, loc) }
+
+      case Expression.Resume(exp, tpe, loc) =>
+        val expVal = visit(exp)
+        mapN(expVal) { e => Expression.Resume(e, tpe, loc) }
+
+      case Expression.InvokeConstructor(constructor, args, tpe, pur, eff, loc) =>
+        val argsVal = traverse(args)(visit)
+        mapN(argsVal) { as => Expression.InvokeConstructor(constructor, as, tpe, pur, eff, loc) }
+
+      case Expression.InvokeMethod(method, exp, args, tpe, pur, eff, loc) =>
+        val expVal = visit(exp)
+        val argsVal = traverse(args)(visit)
+        mapN(expVal, argsVal) { (e, as) => Expression.InvokeMethod(method, e, as, tpe, pur, eff, loc) }
+
+      case Expression.InvokeStaticMethod(method, args, tpe, pur, eff, loc) =>
+        val argsVal = traverse(args)(visit)
+        mapN(argsVal) { as => Expression.InvokeStaticMethod(method, as, tpe, pur, eff, loc) }
+
+      case Expression.GetField(field, exp, tpe, pur, eff, loc) =>
+        val expVal = visit(exp)
+        mapN(expVal) { e => Expression.GetField(field, e, tpe, pur, eff, loc) }
+
+      case Expression.PutField(field, exp1, exp2, tpe, pur, eff, loc) =>
+        val expVal1 = visit(exp1)
+        val expVal2 = visit(exp2)
+        mapN(expVal1, expVal2) { (e1, e2) => Expression.PutField(field, e1, e2, tpe, pur, eff, loc) }
+
+      case Expression.GetStaticField(field, tpe, pur, eff, loc) =>
+        Expression.GetStaticField(field, tpe, pur, eff, loc).toSuccess
+
+      case Expression.PutStaticField(field, exp, tpe, pur, eff, loc) =>
+        val expVal = visit(exp)
+        mapN(expVal) { e => Expression.PutStaticField(field, e, tpe, pur, eff, loc) }
+
+      case Expression.NewObject(name, clazz, tpe, pur, eff, methods, loc) =>
         val erasedType = Type.eraseAliases(tpe)
-        checkObjectImplementation(clazz, erasedType, methods, loc) ++
-          methods.flatMap {
-            case JvmMethod(_, _, exp, _, _, _, _) => visit(exp)
-          }
-
-      case Expression.NewChannel(exp1, exp2, _, _, _, _) =>
-        visit(exp1) ::: visit(exp2)
-
-      case Expression.GetChannel(exp, _, _, _, _) =>
-        visit(exp)
-
-      case Expression.PutChannel(exp1, exp2, _, _, _, _) =>
-        visit(exp1) ::: visit(exp2)
-
-      case Expression.SelectChannel(rules, default, _, _, _, _) =>
-        rules.flatMap { case SelectChannelRule(_, chan, body) => visit(chan) :::
-          visit(body)
-        } :::
-          default.map(visit).getOrElse(Nil)
+        val objImpl = traverse(checkObjectImplementation(clazz, erasedType, methods, loc)) { x => x.toFailure }
```

Comment:
This looks strange to me. Perhaps `checkObjectImplementation` should just return a `Validation[ObjectImplementationOrWhatever, SafetyError]`?

Reply:
Not gonna disagree. Mostly a result of following the old structure.